protocol/x/affiliates/keeper/keeper.go

```diff
@@ -41,3 +48,177 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
+		referredVolume = prevReferredVolumeFromState.BigInt()
+	}
+
+	referredVolume.Add(
+		referredVolume,
+		referredVolumeFromBlock,
+	)
+	updatedReferedVolume := dtypes.NewIntFromBigInt(referredVolume)
+
+	updatedReferredVolumeBytes, err := updatedReferedVolume.Marshal()
+	if err != nil {
+		return errorsmod.Wrapf(types.ErrUpdatingAffiliateReferredVolume,
+			"affiliate %s, error: %s", affiliateAddr, err)
+	}
+	affiliateReferredVolumePrefixStore.Set([]byte(affiliateAddr), updatedReferredVolumeBytes)
+	return nil
+}
+
+// GetReferredVolume returns all time referred volume for an affiliate address.
+func (k Keeper) GetReferredVolume(ctx sdk.Context, affiliateAddr string) (*big.Int, error) {
+	affiliateReferredVolumePrefixStore := prefix.NewStore(ctx.KVStore(k.storeKey), []byte(types.ReferredVolumeKeyPrefix))
+	if !affiliateReferredVolumePrefixStore.Has([]byte(affiliateAddr)) {
+		return big.NewInt(0), nil
+	}
+	var referredVolume dtypes.SerializableInt
+	if err := referredVolume.Unmarshal(affiliateReferredVolumePrefixStore.Get([]byte(affiliateAddr))); err != nil {
+		return big.NewInt(0), err
+	}
+	return referredVolume.BigInt(), nil
+}
+
+// GetAllAffiliateTiers returns all affiliate tiers.
+func (k Keeper) GetAllAffiliateTiers(ctx sdk.Context) (types.AffiliateTiers, error) {
+	store := ctx.KVStore(k.storeKey)
+	affiliateTiersBytes := store.Get([]byte(types.AffiliateTiersKey))
+
+	var affiliateTiers types.AffiliateTiers
+	if affiliateTiersBytes == nil {
+		return affiliateTiers, errorsmod.Wrapf(types.ErrAffiliateTiersNotInitialized, "affiliate tiers not initialized")
+	}
+	err := k.cdc.Unmarshal(affiliateTiersBytes, &affiliateTiers)
+	if err != nil {
+		return affiliateTiers, err
+	}
+
+	return affiliateTiers, nil
+}
+
+// GetTakerFeeShare returns the taker fee share for an address.
+func (k Keeper) GetTakerFeeShare(
+	ctx sdk.Context,
+	address string,
+) (
+	affiliateAddress string,
+	feeSharePpm uint32,
+	exists bool,
+	err error,
+) {
+	affiliateAddress, exists = k.GetReferredBy(ctx, address)
+	if !exists {
+		return "", 0, false, nil
+	}
+	_, feeSharePpm, err = k.GetTierForAffiliate(ctx, affiliateAddress)
+	if err != nil {
+		return "", 0, false, err
+	}
+	return affiliateAddress, feeSharePpm, true, nil
+}
+
+// GetTierForAffiliate returns the tier an affiliate address is qualified for.
+// Assumes that the affiliate tiers are sorted by level in ascending order.
+func (k Keeper) GetTierForAffiliate(
+	ctx sdk.Context,
+	affiliateAddr string,
+) (
+	tierLevel uint32,
+	feeSharePpm uint32,
+	err error) {
+	affiliateTiers, err := k.GetAllAffiliateTiers(ctx)
+	if err != nil {
+		return 0, 0, err
+	}
+	tiers := affiliateTiers.GetTiers()
+	numTiers := uint32(len(tiers))
+	maxTierLevel := numTiers - 1
+	currentTier := uint32(0)
+	referredVolume, err := k.GetReferredVolume(ctx, affiliateAddr)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	for index, tier := range tiers {
+		if referredVolume.Cmp(lib.BigU(tier.ReqReferredVolumeQuoteQuantums)) >= 0 {
```

Comment:
Break early? 

```
	for index, tier := range tiers {
		if referredVolume.Cmp(lib.BigU(tier.ReqReferredVolumeQuoteQuantums)) < 0 {
			break
		}
		// safe to do as tier cannot be negative
		currentTier = uint32(index)
	}
	```

Reply:
Let's also add comment about assumptions on `referredVolume`, `ReqStakedWholeCoins` being strictly increasing in the tiers

Reply:
Did we end up adding comment for this? Would be helpful for both loops